semgrep-core/src/matching/Matching_generic.ml

```diff
@@ -575,6 +575,79 @@ let rec m_list_in_any_order ~less_is_ok f xsa xsb =
       in
       aux candidates
 
+(* ---------------------------------------------------------------------- *)
+(* stdlib: combinatorial search *)
+(* ---------------------------------------------------------------------- *)
+
+(* These functions are not regulat `tin -> tout` matchers, but "combinatorial"
+ * matchers `tin -> ('a list, tout) list. These `m_combs_xyz tin` return a list
+ * of `(rest, tout)` pairs, where `rest` is list of elements remaining to be
+ * matched.
+ *
+ * Used for Associative-Commutative (AC) matching! *)
+
+(* unit operation for the ('a list * tout) list monad *)
+let m_combs_unit xs (tin : tin) : (_ list * tout) list = [ (xs, [ tin ]) ]
+
+(* bind operation for the ('a list * tout) list monad *)
+let rec m_combs_bind comb_matches f : (_ list * tout) list =
+  match comb_matches with
+  | [] -> []
+  | (bs, tout) :: comb_matches' ->
+      let bs_matches = tout |> List.map (fun tin -> f bs tin) |> List.flatten in
+      bs_matches @ m_combs_bind comb_matches' f
+
+let m_combs_flatten comb_result (tin : tin) : tout =
+  comb_result tin |> List.map snd |> List.flatten
+
+let m_combs_fold m_combs xs comb_matches : (_ list * tout) list =
+  List.fold_left
+    (fun combs x -> m_combs_bind combs (fun ys' tin -> m_combs x ys' tin))
+    comb_matches xs
+
+(* Tries matching `a` against each `b_i` in `bs`, and for each succesful match
+ * returns `(bs \ b_i, m a b_i)`. This is essentially the combinatorial version
+ * of `or_list`. *)
+let m_combs_or (m : _ matcher) a bs tin : (_ list * tout) list =
```

Comment:
hmm, we should probably use that in the current implementation of all_elem_and_rest_of_list?

Reply:
Yes, I can do the refactoring!